packages/discovery-provider/src/tasks/index_payment_router.py

```diff
@@ -333,60 +358,232 @@ def index_purchase(
             slot=slot,
             timestamp=timestamp,
             tx_sig=tx_sig,
         )
     # For invalid purchases or transfers not related to a purchase, we'll index
-    # it as a regular transfer, though it will always show as being sent from
-    # the payment router PDA
-    # TODO: We _could_ receive the actual sender from the first tranfer instruction here
-    # and use that instead.
-    else:
-        # TODO: Detect transfers _out_ of user banks to payment router and index them
-        # here as "sent" transactions
-        for user_account in receiver_user_accounts:
-            balance_change = balance_changes[user_account["user_bank_account"]]
-            usdc_tx_received = USDCTransactionsHistory(
-                user_bank=user_account["user_bank_account"],
-                slot=slot,
-                signature=tx_sig,
-                transaction_type=USDCTransactionType.transfer,
-                method=USDCTransactionMethod.receive,
-                transaction_created_at=timestamp,
-                change=Decimal(balance_change["change"]),
-                balance=Decimal(balance_change["post_balance"]),
-                tx_metadata=sender_account,
+    # it as a regular transfer from the sender_account.
+    elif memo["type"] is RouteTransactionMemoType.recovery:
+        try:
+            attempt_index_recovery_transfer(
+                session=session,
+                sender_account=sender_account,
+                sender_user_account=sender_user_account,
+                receiver_user_accounts=receiver_user_accounts,
+                balance_changes=balance_changes,
+                tx_sig=tx_sig,
             )
-            session.add(usdc_tx_received)
-            logger.debug(
-                f"index_payment_router.py | Creating transfer received tx {usdc_tx_received}"
+        except Exception as e:
+            logger.warn(
+                f"index_payment_router.py | tx: {tx_sig} | Failed to index recovery transfer. Will index as plain transfer. Exception received was: {e}."
             )
+            index_transfer(
+                session=session,
+                sender_account=sender_account,
+                sender_user_account=sender_user_account,
+                receiver_user_accounts=receiver_user_accounts,
+                receiver_accounts=receiver_accounts,
+                balance_changes=balance_changes,
+                slot=slot,
+                timestamp=timestamp,
+                tx_sig=tx_sig,
+            )
+    else:
+        index_transfer(
+            session=session,
+            sender_account=sender_account,
+            sender_user_account=sender_user_account,
+            receiver_user_accounts=receiver_user_accounts,
+            receiver_accounts=receiver_accounts,
+            balance_changes=balance_changes,
+            slot=slot,
+            timestamp=timestamp,
+            tx_sig=tx_sig,
+        )
+
+
+def find_sender_account_from_balance_changes(
```

Comment:
this seems a tad scary but i know you've had a bunch of conversations about this. just wanna make sure we have full precision here and there's no case where the total balance wouldn't equal the sum of all the balance_changes.

Reply:
This is a good question. The most correct way to do this would be to correctly find and decode the transfer instruction _into_ payment router and read the source address. Do to the way the indexers currently fetch data, I don't think we have a reliable way to do that without a broad change. @rickyrombo and I talked about it at length and settled on this being a reasonable approach. 
This relies heavily on the client following some sensible rules about transactions:
* We don't send more money into payment router than we are actually routing to the recipients. This would be silly anyway as it leaves funds there which could be arbitrarily sent to any account after the transaction runes
* We don't create transactions which send funds both into payment router and to other addresses via separate instructions.
* We don't create transactions where the source address is also one of the recipients in the `route` instruction.

All of those would cause this check to fail. But those are all cases we _shouldn't_ be exercising in the client. And ultimately, if we fail to detect the sender, then we would just fail to process a recovery transaction and/or an external transfer from a user bank. So it would be a minor annoyance, but wouldn't break any crucial functionality such as content purchase (which uses the memo)

Reply:
That would be a little less prone to failure. Though you'd still end up with the wrong sender if you made a transaction that sent a small amount into payment router and dispersed it and then also sent another amount from a different account to a different destination. Probably even less likely than the cases listed above though!

Reply:
I updated this to take the sender with the greatest loss. Tests still pass. Yay tests.